### PR TITLE
Initial implementation of MP Rest Client 3.0 / MP Fault Tolerance 4.0

### PR DESCRIPTION
**Pull request by andymc12**
Implementation code for MP Fault Tolerance integration with MP Rest Client 3.0 - including updated FATs. Note that this is incomplete - the sync timeout test is failing as is app-supplied interceptors.

### CODE REVIEW

**Comment by andymc12** [comment]
#build

File: dev/build.image/scripts/Liberty- server remote debug.launch [hunk 1]
```diff
@@ -7,14 +7,15 @@
...
... standalone=&amp;quot;no&amp;quot;?&amp;gt;&amp;#10;&amp;lt;javaProject name=&amp;quot;com.ibm.ws.microprofile.health.2.0_fat_tck&amp;quot;/&amp;gt;&amp;#10;&quot; typeId=&quot;org.eclipse.jdt.launching.sourceContainer.javaProject&quot;/&gt;&#10;        &lt;container memento=&quot;&amp;lt;?xml version=&amp;quot;1.0&amp;quot; encoding=&amp;quot;UTF-8&amp;quot; standalone=&amp;quot;no&amp;quot;?&amp;gt;&amp;#10;&amp;lt;javaProject name=&amp;quot;com.ibm.ws.microprofile.health.2.1_fat_tck&amp;quot;/&amp;gt;&amp;#10;&quot; typeId=&quot;org.eclipse.jdt.launching.sourceContainer.javaProject&quot;/&gt;&#10;        &lt;container memento=&quot;&amp;lt;?xml version=&amp;quot;1.0&amp;quot; encoding=&amp;quot;UTF-8&amp;quot; standalone=&amp;quot;no&amp;quot;?&amp;gt;&amp;#10;&amp;lt;javaProject name=&amp;quot;com.ibm.ws.microprofile.health.2.2_fat_tck&amp;quot;/&amp;gt;&amp;#10;&quot; typeId=&quot;org.eclipse.jdt.launching.sourceContainer.javaProject&quot;/&gt;&#10;        &lt;container memento=&quot;&amp;lt;?xml version=&amp;quot;1.0&amp;quot; encoding=&amp;quot;UTF-8&amp;quot; standalone=&amp;quot;no&amp;quot;?&amp;gt;&amp;#10;&amp;lt;javaProject name=&amp;quot;com.ibm.ws.microprofile.metric
```
Comment (tevans78) [inline dev/build.image/scripts/Liberty- server remote debug.launch:10]: This file probably shouldn't be included in this PR? Shouldn't break anything but it isn't relevant

File: dev/com.ibm.ws.microprofile.rest.client.FT_fat/fat/src/com/ibm/ws/microprofile/rest/client/fat/FATSuite.java [hunk 2]
```diff
@@ -17,49 +17,43 @@
...
             case "1.2":
-            case "1.3": return use(action, "mpConfig", "1.3", "1.0", "1.1", "1.2", "1.4", "2.0");
-            case "1.4": return use(action, "mpConfig", "1.4", "1.0", "1.1", "1.2", "1.3", "2.0");
-            case "2.0":
-            default:    return use(action, "mpConfig", "2.0", "1.0", "1.1", "1.2", "1.3", "1.4");
+            case "1.3": return use(action, "mpConfig", "1.3", "1.0", "1.1", "1.2", "1.4", "2.0", "3.0");
+            case "1.4": return use(action, "mpConfig", "1.4", "1.0", "1.1", "1.2", "1.3", "2.0", "3.0");
+            case "2.0": return use(action, "mpConfig", "2.0", "1.0", "1.1", "1.2", "1.3", "1.4", "3.0");
+            case "3.0":
+            default:    return use(action, "mpConfig", "3.0", "1.0", "1.1", "1.2", "1.3", "1.4", "2.0");
```
Comment (tevans78) [inline dev/com.ibm.ws.microprofile.rest.client.FT_fat/fat/src/com/ibm/ws/microprofile/rest/client/fat/FATSuite.java:40]: This isn't wrong but I would have recommended using the MicroProfileActions class

File: dev/com.ibm.ws.microprofile.rest.client.FT_fat/fat/src/com/ibm/ws/microprofile/rest/client/fat/RetryTest.java [hunk 3]
```diff
@@ -49,7 +53,11 @@
 
     @BeforeClass
     public static void setUp() throws Exception {
-        ShrinkHelper.defaultDropinApp(server, appName, SERVER_NAME);
+        WebArchive war = ShrinkHelper.buildDefaultApp(appName, SERVER_NAME);
+        StringAsset mpConfig = new StringAsset(RetryClient.class.getName() + "/mp-rest/uri=http://localhost:"
+                        + System.getProperty("bvt.prop.HTTP_default", "8010") + "/retryApp");
+        war.addAsWebInfResource(mpConfig, "classes/META-INF/microprofile-config.properties");
+        ShrinkHelper.exportDropinAppToServer(server, war);
```
Comment (tevans78) [inline dev/com.ibm.ws.microprofile.rest.client.FT_fat/fat/src/com/ibm/ws/microprofile/rest/client/fat/RetryTest.java]: recommend adding `DeployOptions.SERVER_ONLY` otherwise there can be contamination between repeats, particularly when mixing EE8 and EE9

**Comment by andymc12** [comment]
Fixed a null check issue and manually kicked off build: https://wasrtc.hursley.ibm.com:9443/jazz/web/projects/WS-CD#action=com.ibm.team.build.viewResult&id=_DeZSUFgyEeyWaZwSyDrk3A

Just saw Tom's code review comments - will address these now.

File: dev/com.ibm.ws.microprofile.rest.client.FT_fat/test-applications/retryApp/resources/WEB-INF/beans.xml [hunk 4]
```diff
@@ -1,9 +1,4 @@
-<beans
-   xmlns="http://java.sun.com/xml/ns/javaee"
-   xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
-   xsi:schemaLocation="
-      http://java.sun.com/xml/ns/javaee
-      http://java.sun.com/xml/ns/javaee/beans_1_0.xsd">
+<beans>
```
Comment (tevans78) [inline dev/com.ibm.ws.microprofile.rest.client.FT_fat/test-applications/retryApp/resources/WEB-INF/beans.xml:1]: Strictly neither the old or new versions are valid beans.xml
```suggestion
<?xml version="1.0" encoding="UTF-8"?>
<beans xmlns="http://java.sun.com/xml/ns/javaee"
       xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
       xsi:schemaLocation="http://java.sun.com/xml/ns/javaee http://java.sun.com/xml/ns/javaee/beans_1_0.xsd">
```

File: dev/com.ibm.ws.microprofile.rest.client.FT_fat/test-applications/retryApp/src/mpRestClientFT/retry/RetryTestServlet.java [hunk 5]
```diff
@@ -63,6 +59,7 @@ public void testRetryOnceOnFailThenSucceed_separateMethod(HttpServletRequest req
     }
 
     @Test
+    @SkipForRepeat(JakartaEE9Action.ID) // CDI not resolving app-supplied interceptor here TODO: investigate and resolve
```
Comment (tevans78) [inline dev/com.ibm.ws.microprofile.rest.client.FT_fat/test-applications/retryApp/src/mpRestClientFT/retry/RetryTestServlet.java:62]: is this a big deal? is there an issue open?

File: dev/build.image/scripts/Liberty- server remote debug.launch (hunk 1 above)
Comment (andymc12) [reply dev/build.image/scripts/Liberty- server remote debug.launch:10]: It adds some of the recently added MP source projects to the debug lookup path.  It definitely helped me identify where the rest client and FT were failing, so I think it would be a good thing to have - I think it could benefit others who might need to debug these same projects in the future.  If you want me to remove the change though, I will. Please let me know.

File: dev/io.openliberty.org.jboss.resteasy.mprestclient/bnd.bnd [hunk 6]
```diff
@@ -52,6 +52,9 @@ Import-Package: \
   org.eclipse.microprofile.rest.client.*;version='!',\
   *
 
+DynamicImport-Package: \
+  com.ibm.ws.microprofile.faulttolerance.cdi
```
Comment (tevans78) [inline dev/io.openliberty.org.jboss.resteasy.mprestclient/bnd.bnd:56]: I don't believe that this package is intended to be used by other bundles?

File: dev/com.ibm.ws.microprofile.rest.client.FT_fat/test-applications/retryApp/resources/WEB-INF/beans.xml (hunk 4 above)
Comment (andymc12) [reply dev/com.ibm.ws.microprofile.rest.client.FT_fat/test-applications/retryApp/resources/WEB-INF/beans.xml:1]: I based the changes on this article:
https://blog.payara.fish/getting-started-with-jakarta-ee-9-context-and-dependency-injection-cdi

In the "interceptors" section, Rudy De Busscher writes:
```
<beans>
   <interceptors>
      <class>fish.payara.jakarta.ee9.start.TimedInterceptor</class>
    </interceptors>
</beans>
```
> In the above example, we did not specify the namespace of the beans tag. This has the advantage that it can be used by different versions of Java EE and Jakarta EE (since the namespace changes between Java EE, Jakarta EE 8, and Jakarta EE 9) but we do not specify the bean discovery mode.

Please let me know if there is a better approach.

File: dev/com.ibm.ws.microprofile.rest.client.FT_fat/test-applications/retryApp/src/mpRestClientFT/retry/RetryTestServlet.java (hunk 5 above)
Comment (andymc12) [reply dev/com.ibm.ws.microprofile.rest.client.FT_fat/test-applications/retryApp/src/mpRestClientFT/retry/RetryTestServlet.java:62]: Open issue - possibly related to the beans.xml issue you pointed out earlier??

File: dev/io.openliberty.org.jboss.resteasy.mprestclient/bnd.bnd (hunk 6 above)
Comment (andymc12) [reply dev/io.openliberty.org.jboss.resteasy.mprestclient/bnd.bnd:56]: It's necessary in order for the rest client implementation to add the Liberty-specific FT annotation class.  Since the actual "bean" is a proxy class, the existing FT code doesn't handle the annotations.  We did the same thing with MP Rest Client 1.X and 2.X.

File: dev/io.openliberty.org.jboss.resteasy.mprestclient/src/org/jboss/resteasy/microprofile/client/ot/LibertyRestClientBuilderImpl.java [hunk 7]
```diff
@@ -116,12 +127,30 @@
...
+            });
+        } catch (PrivilegedActionException pae) {
+            if (LOGGER.isDebugEnabled()) {
+                LOGGER.debug("Exception checking for MP Fault Tolerance class - " +
+                             "expected if FT feature isnot enabled", pae);
+            }
+            return null;
+        }
+    }
+
```
Comment (Azquelt) [inline dev/io.openliberty.org.jboss.resteasy.mprestclient/src/org/jboss/resteasy/microprofile/client/ot/LibertyRestClientBuilderImpl.java:153]: We should probably have a proper integration point rather than just looking to see if an internal class is loadable.

I don't understand enough about what else is going on in this class to make a sensible suggestion about what it should be though. Could we have a chat about the best way to do this?

File: dev/com.ibm.ws.microprofile.rest.client.FT_fat/test-applications/retryApp/resources/WEB-INF/beans.xml (hunk 4 above)
Comment (tevans78) [reply dev/com.ibm.ws.microprofile.rest.client.FT_fat/test-applications/retryApp/resources/WEB-INF/beans.xml:1]: I suspect that article was really just using snippets. Without the xml header and namespaces etc, it might still work but it is not _strictly_ correct. Whether is works or not would be implementation specific.

Comment (tevans78) [reply dev/com.ibm.ws.microprofile.rest.client.FT_fat/test-applications/retryApp/resources/WEB-INF/beans.xml:1]: If you are seeing other issues then is is possible that the file is just being silently ignored??

File: dev/com.ibm.ws.microprofile.rest.client.FT_fat/test-applications/retryApp/src/mpRestClientFT/retry/RetryTestServlet.java (hunk 5 above)
Comment (tevans78) [reply dev/com.ibm.ws.microprofile.rest.client.FT_fat/test-applications/retryApp/src/mpRestClientFT/retry/RetryTestServlet.java:62]: please can you add a link to the open issue? here and/or in the code

**Review by tevans78 — APPROVED** [review]
While there are some issues in this PR, I think they can be addressed later. In particular we should look at how this (and the previous version) uses our internal `com.ibm.ws.microprofile.faulttolerance.cdi.FaultTolerance` annotation class.

**Comment by andymc12** [comment]
New #build to include latest changes (DeployOptions in FAT tests)

**Comment by tevans78** [comment]
Failures in PB look unrelated. Merging.

**Comment by tevans78** [comment]
#libby
